File: Onboarding.swift
import SwiftUI

struct VocaliaStartView: View {
    @AppStorage("languageChoice102Release") private var completed = false
    var body: some View {
        if completed { TranscribeView() }
        else { VocaliaLanguageChoice { completed = true } }
    }
}

struct VocaliaLanguageChoice: View {
    @AppStorage("uiLanguage") private var language = AppLanguage.system
    var onFinish: () -> Void
    var body: some View {
        VStack(alignment: .leading, spacing: 24) {
            Text("Vocalia").font(.system(size: 32, weight: .bold, design: .rounded))
            Text(T("Welcome")).font(.title2.bold())
            Text(T("Choose your language before starting.")).foregroundStyle(.secondary)
            Picker(T("App language"), selection: $language) {
                ForEach(AppLanguage.options,id:\.code) { option in Text(option.name).tag(option.code) }
            }.pickerStyle(.menu)
            HStack { Spacer(); Button(T("Continue"), action: onFinish).buttonStyle(.borderedProminent).keyboardShortcut(.defaultAction) }
        }.padding(40).frame(width: 480)
    }
}

struct TourAnchors: PreferenceKey {
    static var defaultValue: [String: Anchor<CGRect>] { [:] }
    static func reduce(value: inout [String: Anchor<CGRect>], nextValue: () -> [String: Anchor<CGRect>]) {
        value.merge(nextValue(), uniquingKeysWith: { _, new in new })
    }
}
extension View {
    func tourTarget(_ key: String) -> some View {
        transformAnchorPreference(key: TourAnchors.self, value: .bounds) { values, anchor in values[key] = anchor }
    }
}

struct VocaliaCoachMarks: View {
    var anchors: [String: Anchor<CGRect>]
    var language: String
    var onFinish: () -> Void
    @State private var index = 0
    private func text(_ value:String)->String {AppLanguage.interface(value,language:language)}
    private let steps: [(String,String,String,String,String)] = [
        ("add", "Agrega tus grabaciones", "Agrega o arrastra grabaciones para iniciar la transcripción automáticamente. Los archivos se procesan uno a uno, incluidos OPUS.", "Add your recordings", "Add or drop recordings to start transcription automatically. Multiple files are processed one at a time, including OPUS."),
        ("audio", "Idioma de la grabación", "Elige el idioma que habla la persona. Este ajuste es independiente del idioma de los botones.", "Recording language", "Choose the audio language before adding recordings. It is independent of the interface language."),
        ("prepare", "Prepara el reconocimiento", "Si el idioma no está listo, descárgalo aquí con internet. Después puedes transcribir sin conexión. Tus grabaciones se quedan en tu Mac.", "Prepare speech recognition", "If a model is missing, download it once; transcription continues when it is ready. Recordings stay on this computer."),
        ("run", "Retoma grabaciones detenidas", "Los archivos nuevos empiezan automáticamente. Este botón reintenta grabaciones pendientes o detenidas. Cancelar detiene la cola; el análisis de voces sigue siendo manual.", "Resume stopped recordings", "New files start automatically. This button retries pending or stopped recordings. Cancel stops the queue; speaker analysis remains manual."),
        ("status", "Sigue el progreso", "Aquí verás la preparación, el progreso y el estado del reconocimiento. Espera a que termine para copiar el resultado completo.", "Follow progress", "This area shows preparation, progress and recognition status. Wait for completion before copying the full result."),
        ("editor", "Escucha, revisa y copia", "Aquí aparecerá tu transcripción completa. Revisa nombres y cifras. Reproducir solo sirve para escuchar; el resaltado sigue el audio. Usa Copiar todo el texto o Exportar para llevarte el resultado y selecciona frases para guardar cuñas.", "Listen, review and copy", "Your full transcript appears here. Check names and numbers. Play only plays audio; highlighting follows the recording. Use Copy full text or Export to take the result with you, and select passages to save quotes."),
        ("interface", "Cambia el idioma cuando quieras", "Aquí cambias los botones entre español e inglés. Puedes repetir este recorrido desde Cómo usar Vocalia.", "Change language any time", "Choose one of the six interface languages here. Reopen this tour from How to use Vocalia.")
    ]
    var body: some View {
        GeometryReader { proxy in
            let step = steps[index]
            if let anchor = anchors[step.0] {
                let target = proxy[anchor]
                let width: CGFloat = 350
                let height: CGFloat = 290
                let right = proxy.size.width - target.maxX > width + 24
                let left = target.minX > width + 24
                let x = min(max(right ? target.maxX + 18 : left ? target.minX - width - 18 : target.midX - width/2, 12), proxy.size.width - width - 12)
                let proposedY = right || left ? target.midY - height/2 : target.maxY + height + 24 < proxy.size.height ? target.maxY + 18 : target.minY - height - 18
                let y = min(max(proposedY, 12), proxy.size.height-height-12)
                RoundedRectangle(cornerRadius: 8).stroke(.indigo, lineWidth: 3)
                    .frame(width: max(0,target.width+8), height: max(0,target.height+8))
                    .position(x:target.midX,y:target.midY).allowsHitTesting(false)
                VStack(alignment: .leading, spacing: 14) {
                    HStack {
                        Text("\(index+1) / \(steps.count)").font(.caption).foregroundStyle(.secondary)
                        Spacer()
                        Button(text("Skip"), action:onFinish).buttonStyle(.plain).keyboardShortcut(.cancelAction)
                    }
                    Text(text(step.3)).font(.headline)
                    ScrollView { Text(text(step.4)).font(.system(size:14)).frame(maxWidth:.infinity,alignment:.leading).fixedSize(horizontal:false,vertical:true) }
                    HStack {
                        Button(text("Back")) { index -= 1 }.disabled(index == 0)
                        Spacer()
                        Button(index == steps.count-1 ? text("Done") : text("Next")) {
                            if index == steps.count-1 { onFinish() } else { index += 1 }
                        }.buttonStyle(.borderedProminent).keyboardShortcut(.defaultAction)
                    }
                }.padding(18).frame(width:width,height:height)
                    .background(Color.white,in:RoundedRectangle(cornerRadius:16))
                    .overlay(RoundedRectangle(cornerRadius:16).stroke(.indigo.opacity(0.6)))
                    .shadow(color:.black.opacity(0.15),radius:12,y:4)
                    .position(x:x+width/2,y:y+height/2)
                Text(right ? "◀" : left ? "▶" : y > target.midY ? "▲" : "▼")
                    .font(.system(size:22)).foregroundStyle(.indigo)
                    .position(x: right ? x-6 : left ? x+width+6 : min(max(target.midX,x+20),x+width-20),
                              y: right || left ? min(max(target.midY,y+24),y+height-24) : y > target.midY ? y-7 : y+height+7)
                    .allowsHitTesting(false)
            }
        }
    }
}

File: Vocalia-Windows/onboarding.py
"""Local language selection and anchored, replayable in-app coach marks."""
from PySide6.QtCore import Qt, QTimer, QPoint
from PySide6.QtWidgets import (QDialog, QVBoxLayout, QLabel, QComboBox,
    QPushButton, QHBoxLayout, QFrame, QRubberBand)

from i18n import LANGUAGES, tr, normalize

class WelcomeDialog(QDialog):
    def __init__(self, language='en', parent=None):
        super().__init__(parent)
        self.setWindowTitle('Vocalia'); self.setMinimumWidth(420)
        self.language = language
        layout = QVBoxLayout(self); layout.setSpacing(20)
        self.heading = QLabel();self.heading.setWordWrap(True);self.heading.setStyleSheet('font-size:20px;font-weight:700');layout.addWidget(self.heading)
        self.description=QLabel();self.description.setWordWrap(True);layout.addWidget(self.description)
        self.language_box = QComboBox()
        for code,name in LANGUAGES:self.language_box.addItem(name,code)
        self.language_box.setCurrentIndex(self.language_box.findData(normalize(language)))
        layout.addWidget(self.language_box)
        self.next_button = QPushButton()
        self.next_button.clicked.connect(self.advance); layout.addWidget(self.next_button)
        self.language_box.currentIndexChanged.connect(self.render)
        self.render()
    def render(self):
        language=self.language_box.currentData()
        self.heading.setText(tr('Welcome to Vocalia',language));self.description.setText(tr('Choose your language before starting.',language));self.next_button.setText(tr('Continue',language))
    def advance(self):
        self.language = self.language_box.currentData(); self.accept()

# key, Spanish title/body, English title/body
STEPS = [
 ('add', 'Agrega tus grabaciones', 'Agrega o arrastra grabaciones para iniciar la transcripción automáticamente. Los archivos se procesan uno a uno, incluidos OPUS.', 'Add your recordings', 'Add or drop recordings to start transcription automatically. Multiple files are processed one at a time, including OPUS.'),
 ('audio', 'Idioma de la grabación', 'Elige el idioma que habla la persona. Este ajuste es independiente del idioma de los botones.', 'Recording language', 'Choose the audio language before adding recordings. It is independent of the interface language.'),
 ('prepare', 'Prepara el reconocimiento', 'Descarga el modelo una vez con internet. Después puedes transcribir sin conexión. Tus grabaciones se quedan en este equipo.', 'Prepare speech recognition', 'If a model is missing, download it once; transcription continues when it is ready. Recordings stay on this computer.'),
 ('run', 'Retoma grabaciones detenidas', 'Los archivos nuevos empiezan automáticamente. Este botón reintenta grabaciones pendientes o detenidas. Cancelar detiene la cola; el análisis de voces sigue siendo manual.', 'Resume stopped recordings', 'New files start automatically. This button retries pending or stopped recordings. Cancel stops the queue; speaker analysis remains manual.'),
 ('status', 'Sigue el progreso', 'Aquí verás si Vocalia está preparando el modelo, transcribiendo o ha encontrado un problema. Espera a que termine para copiar el resultado completo.', 'Follow progress', 'This area shows preparation, transcription and any problems. Wait for completion before copying the full result.'),
 ('play', 'Escucha y revisa', 'Reproducir solo permite escuchar; no transcribe. Cuando haya texto, Seguir audio resaltará la posición mientras escuchas.', 'Listen and review', 'Play only plays the recording; it does not transcribe. Once text is available, Follow audio highlights your position as you listen.'),
 ('tabs', 'Texto completo y cuñas', 'Aquí aparecerá el texto editable. Revisa nombres y cifras. Selecciona una frase y usa Guardar cuña para conservar una cita.', 'Full text and quotes', 'Your editable transcript appears here. Check names and numbers. Select a passage and use Save quote to keep a quote.'),
 ('copy', 'Copia o exporta', 'Copia todo el texto de una vez. El botón Exportar también permite guardar el resultado para trabajar fuera de Vocalia.', 'Copy or export', 'Copy the full text in one go. Export also saves the result for use outside Vocalia.'),
 ('interface', 'Cambia el idioma cuando quieras', 'Aquí cambias los botones entre español e inglés. Puedes repetir este recorrido desde Cómo usar Vocalia.', 'Change language any time', 'Choose one of the six interface languages here. Reopen this tour from How to use Vocalia.')
]

class GuidedTour(QFrame):
    def __init__(self, parent, targets, language, finished):
        super().__init__(parent)
        self.targets = targets; self.language = language; self.finished = finished; self.index = 0
        self.setObjectName('tourBubble')
        self.setStyleSheet('#tourBubble{background:#fff;border:2px solid #6960ad;border-radius:14px} QLabel{color:#292654;border:0} QPushButton{padding:7px}')
        self.setFixedWidth(350)
        layout = QVBoxLayout(self); layout.setContentsMargins(18,16,18,16); layout.setSpacing(12)
        self.counter = QLabel(); layout.addWidget(self.counter)
        self.heading = QLabel(); self.heading.setWordWrap(True); self.heading.setStyleSheet('font-size:18px;font-weight:700');layout.addWidget(self.heading)
        self.detail = QLabel();self.detail.setWordWrap(True);layout.addWidget(self.detail)
        row = QHBoxLayout(); layout.addLayout(row)
        self.skip = QPushButton(tr('Skip',language)); row.addWidget(self.skip)
        row.addStretch()
        self.back = QPushButton(tr('Back',language));row.addWidget(self.back)
        self.next = QPushButton();row.addWidget(self.next)
        self.skip.clicked.connect(self.finish); self.back.clicked.connect(lambda:self.go(-1)); self.next.clicked.connect(lambda:self.go(1))
        self.ring = QRubberBand(QRubberBand.Shape.Rectangle, parent)
        self.ring.setAttribute(Qt.WidgetAttribute.WA_TransparentForMouseEvents)
        self.arrow = QLabel(parent);self.arrow.setStyleSheet('color:#6960ad;font-size:22px;background:transparent')
        self.arrow.setAttribute(Qt.WidgetAttribute.WA_TransparentForMouseEvents)
        self.timer = QTimer(self);self.timer.setInterval(120);self.timer.timeout.connect(self.place)
        self.render();self.show();self.timer.start()
    def render(self):
        step = STEPS[self.index]
        self.counter.setText(f'{self.index+1} / {len(STEPS)}')
        self.heading.setText(tr(step[3],self.language)); self.detail.setText(tr(step[4],self.language))
        self.back.setEnabled(self.index > 0)
        self.next.setText(tr('Done' if self.index == len(STEPS)-1 else 'Next',self.language))
        self.adjustSize();self.place();self.next.setFocus()
        self.setAccessibleName(self.heading.text());self.detail.setAccessibleName(self.detail.text())
    def place(self):
        target = self.targets[STEPS[self.index][0]]
        point = target.mapTo(self.parentWidget(), QPoint(0,0))
        rect = target.rect().translated(point).adjusted(-3,-3,3,3)
        self.ring.setGeometry(rect);self.ring.show();self.ring.raise_()
        bounds = self.parentWidget().rect(); margin=12
        # Prefer a side with enough space, keeping the highlighted control exposed.
        if bounds.width()-rect.right() >= self.width()+24:
            x=rect.right()+18;y=rect.center().y()-self.height()//2;glyph='◀'
        elif rect.left() >= self.width()+24:
            x=rect.left()-self.width()-18;y=rect.center().y()-self.height()//2;glyph='▶'
        else:
            x=rect.center().x()-self.width()//2
            below=bounds.height()-rect.bottom()
            y=rect.bottom()+18 if below>=self.height()+24 else rect.top()-self.height()-18
            glyph='▲' if below>=self.height()+24 else '▼'
        x=max(margin,min(x,bounds.width()-self.width()-margin));y=max(margin,min(y,bounds.height()-self.height()-margin))
        self.move(x,y);self.raise_()
        self.arrow.setText(glyph);self.arrow.adjustSize()
        if glyph in ('◀','▶'):
            ax=x-16 if glyph=='◀' else x+self.width()-2
            ay=max(y+12,min(rect.center().y()-12,y+self.height()-28))
        else:
            ax=max(x+12,min(rect.center().x()-10,x+self.width()-28));ay=y-19 if glyph=='▲' else y+self.height()-3
        self.arrow.move(ax,ay);self.arrow.show();self.arrow.raise_()
    def go(self, delta):
        if self.index+delta >= len(STEPS):self.finish();return
        self.index=max(0,self.index+delta);self.render()
    def finish(self):
        self.timer.stop();self.ring.hide();self.arrow.hide();self.ring.deleteLater();self.arrow.deleteLater();self.hide();self.finished();self.deleteLater()
    def keyPressEvent(self,event):
        if event.key()==Qt.Key.Key_Escape:self.finish()
        else:super().keyPressEvent(event)
